Assets/MixedRealityToolkit.SDK/Experimental/Features/Dwell/DwellHandler.cs

```diff
@@ -0,0 +1,210 @@
+        public virtual float CalculateDwellProgress()
+        {
+            float dwellProgress = 0;
+
+            switch (currentDwellState)
+            {
+                case DwellState.None:
+                    dwellProgress = 0;
+                    break;
+                case DwellState.FocusGained:
+                    dwellProgress = 0;
+                    fillTimer = 0;
+                    break;
+                case DwellState.DwellStarted:
+                    fillTimer -= Time.deltaTime;
+                    fillTimer = fillTimer < 0 ? 0 : fillTimer;
+                    dwellProgress = Mathf.Clamp((float)(dwellProfile.TimeToCompleteDwell - fillTimer) / dwellProfile.TimeToCompleteDwell, 0f, 1f);
+                    break;
+                case DwellState.DwellCompleted:
+                    dwellProgress = 1;
+                    break;
+                case DwellState.DwellCanceled:
+                    if (dwellProfile.AllowDwellResume && (DateTime.UtcNow - focusExitTime).TotalSeconds > dwellProfile.TimeToAllowDwellResume)
+                    {
+                        dwellProgress = 0;
+                        fillTimer = 0;
+                        currentDwellState = DwellState.None;
+                    }
+                    else if (dwellProfile.AllowDwellResume)
+                    {
+                        dwellProgress = Mathf.Clamp((float)(dwellProfile.TimeToCompleteDwell - fillTimer) / dwellProfile.TimeToCompleteDwell, 0f, 1f);
+                    }
+                    break;
+                case DwellState.Invalid:
+                default:
+                    return dwellProgress;
+            }
+
+            return dwellProgress;
+        }
+
+        public void OnFocusEnter(FocusEventData eventData)
+        {
+            if (eventData.NewFocusedObject == gameObject
+                && eventData.Pointer.InputSourceParent.SourceType == dwellProfile.DwellPointerType)
+            {
+                HasFocus = true;
+
+                // dwell state machine re-starts
+                if (currentDwellState <= DwellState.DwellIntended
+                    || !dwellProfile.AllowDwellResume || (DateTime.UtcNow - focusExitTime).TotalSeconds > dwellProfile.TimeToAllowDwellResume)
+                {
+                    focusEnterTime = DateTime.UtcNow;
+                    currentDwellState = DwellState.FocusGained;
+                    pointerRef = eventData.Pointer;
+                    fillTimer = 0;
+                }
+                //intent to resume 
+                else if (currentDwellState == DwellState.DwellCanceled && dwellProfile.AllowDwellResume
+                    && pointerRef.InputSourceParent.SourceId == eventData.Pointer.InputSourceParent.SourceId //make sure the returning pointer id is the same
+                    && (DateTime.UtcNow - focusExitTime).TotalSeconds <= dwellProfile.TimeToAllowDwellResume)
+                {
```

Comment:
It seems like you're doing a lot of this logic twice, once in the if clause and again here in reverse.  Consider refactoring to something like this:

if (resume criteria)
{
   // do resume actions here
}
else //maybe restart in all other cases?  Not sure if you need to check anything in this case.
{
   // do restart actions here
}